base/asyncmap.jl

```diff
@@ -62,24 +72,34 @@ function done(itr::AsyncCollector, state::AsyncCollectorState)
 end
 
 function next(itr::AsyncCollector, state::AsyncCollectorState)
-    # Wait if the maximum number of concurrent tasks are already running
+    # Wait if the maximum number of concurrent tasks are already running.
     while isbusy(itr, state)
         wait(state)
     end
 
-    # Get index and mapped function arguments from enumeration iterator
+    # Get index and mapped function arguments from enumeration iterator.
     (i, args), state.enum_state = next(itr.enumerator, state.enum_state)
 
-    # Execute function call and save result asynchronously
+    # Execute function call and save result asynchronously.
     @async begin
-        itr.results[i] = itr.f(args...)
-        state.active_count -= 1
+        try
+            itr.results[i] = itr.f(args...)
+        catch e
+            state.in_error = true
+            notify(state.task_done, e; error=true)
```

Comment:
Fair enough, if you think that avoiding starting one more `@async` task is worth it,
what about changing the `wait` in `next` look like this:
```julia
     while isbusy(itr, state)
          wait(state)
          if state.in_error
              # Stop processing immediately on error.
              return (nothing, state)
          end
     end
```